src/kissmvc.php

```diff
@@ -77,172 +66,171 @@ public function __construct($defaultController = '', $defaultAction = '') {
+            header($_SERVER['SERVER_PROTOCOL'].' 404 Not Found', true, 404);
+        } else if ($e->getCode() >= 400 && $e->getCode() < 600) {
+            header($_SERVER['SERVER_PROTOCOL'].' '.$e->getCode().' '.$e->getMessage(), true, $e->getCode());
         } else {
-            header($_SERVER['SERVER_PROTOCOL'] . ' 500 Internal Server Error', true, 500);
+            header($_SERVER['SERVER_PROTOCOL'].' 500 Internal Server Error', true, 500);
         }
     }
-    
-    private function route_request() {
-        $controllerClassName = ucfirst($this->controller).'Controller';
-        include (APP_PATH . 'controllers/' . $controllerClassName . '.php');
-        if (class_exists($controllerClassName)) {
-            $controller = new $controllerClassName();
-        } else {
-            throw new NotFoundException("Controller " . $this->controller . " not found.");
-        }
-        $controller->setParams($this->params);
-        $controller->setArgs($this->args);
-        $view = $controller->process($this->action);
-        if ($view && $view instanceof View) {
-            $view->dump();
-        }
-    }
-    
-    protected function preprocessParams() {
-        $this->params = $_REQUEST;
-    }
 }
 
-//===============================================================
-// Controller
-//===============================================================
+/**
+ * Base controller, override this class and implement individual actions as methods called "_action".
+ */
 abstract class Controller {
-
     private $params;
-    private $args;
 
     /**
-     * process is called from Mvp class.
+     * Override for any parameter validation/preprocessing.
      */
-    function process($action = '') {
-        $action = '_'.$action;
-        if (!method_exists($this, $action)) {
-            throw new NotFoundException("Function '".$action."' not found in ".get_class($this));
+    function preprocessParams() {
+        if (strpos($_SERVER['HTTP_CONTENT_TYPE'], 'application/json') === 0) {
+            $this->setParams(json_decode(file_get_contents('php://input'), true));
+        } else {
+            $this->setParams($_REQUEST);
         }
-        return call_user_func_array($this->$action, $this->args);
     }
-    
-    function setParams($params = []) {
-        $this->params = $params;
+
+    /**
+     * Default request processing. You may override this method for custom handling.
+     */
+    function process($action, $args) {
+        $method = strtolower($_SERVER['REQUEST_METHOD']).($action ? '_'.$action : '');
+        if (!method_exists($this, $method)) {
+            if (method_exists($this, '_'.$action)) {
+                $method = '_'.$action;
+            } else {
+                throw new NotFoundException('Method \''.$method.'\' nor \'_'.$action.'\' not found in '.get_class($this));
+            }
+        }
+        return call_user_func_array($method, $args);
```

Comment:
Toto funguje? Netreba to volat na objekt ako `call_user_func_array([$this, $method], $args);` ?